gb-cpu/src/address_bus.rs

```diff
@@ -0,0 +1,275 @@
+            ram: Box::new(CharDevice(4)),
+            eram: Box::new(CharDevice(5)),
+            oam: Box::new(CharDevice(6)),
+            io_reg: Box::new(CharDevice(7)),
+            hram: Box::new(CharDevice(8)),
+            ie_reg: Box::new(CharDevice(9)),
+        };
+
+        assert_eq!(addr_bus.write(0x30, 0x11), Ok(()));
+        assert_eq!(addr_bus.write(0x31, 0x8242), Ok(()));
+        assert_eq!(addr_bus.write(0x32, 0xa050), Ok(()));
+        assert_eq!(addr_bus.write(0x33, 0xdf8f), Ok(()));
+        assert_eq!(addr_bus.write(0x34, 0xe006), Ok(()));
+        assert_eq!(addr_bus.write(0x35, 0xfe80), Ok(()));
+        assert_eq!(addr_bus.write(0x36, 0xff70), Ok(()));
+        assert_eq!(addr_bus.write(0x37, 0xff8e), Ok(()));
+        assert_eq!(addr_bus.write(0x38, 0xffff), Ok(()));
+
+        assert_eq!(addr_bus.read(0x10), Ok(0x30));
+        assert_eq!(addr_bus.read(0x8042), Ok(0x31));
+        assert_eq!(addr_bus.read(0xa000), Ok(0x32));
+        assert_eq!(addr_bus.read(0xdfff), Ok(0x33));
+        assert_eq!(addr_bus.read(0xe000), Ok(0x34));
+        assert_eq!(addr_bus.read(0xfe00), Ok(0x35));
+        assert_eq!(addr_bus.read(0xff00), Ok(0x36));
+        assert_eq!(addr_bus.read(0xff80), Ok(0x37));
+        assert_eq!(addr_bus.read(0xffff), Ok(0x38));
+    }
+}
+
+pub struct Iter<'a> {
+    current_address: u16,
+    stop: bool,
+    bus: &'a AddressBus,
+}
+
+impl<'a> Iter<'a> {
+    fn new(bus: &'a AddressBus) -> Self {
+        Self {
+            current_address: 0,
+            stop: false,
+            bus,
+        }
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.stop {
+            let bit = self.bus.read(self.current_address).ok();
+            match self.current_address {
+                0xfea0..=0xfeff => self.current_address = 0xff00,
+                0xffff => self.stop = true,
+                _ => self.current_address += 1,
+            }
+            bit
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    BusError(Address),
+    SegmentationFault(Address),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+/// Address contain the relative and absolute address
+pub struct Address {
+    /// relative address into the current area of the address bus
+    pub relative: u16,
+
+    /// absolute address used in the address bus
+    pub absolute: u16,
+
+    pub area: Area,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Area {
+    Bios,
+    Rom,
+    Vram,
+    ExtRam,
+    Ram,
+    ERam,
+    Oam,
+    IoReg,
+    HighRam,
+    IEReg,
+    Unbound,
+}
+
+impl Address {
+    pub fn new(area: Area, relative_addr: u16, absolute_addr: u16) -> Self {
+        Self {
+            relative: relative_addr,
+            absolute: absolute_addr,
+            area,
+        }
+    }
+
+    /// Create a Address from an absolute adress and an offset
```

Comment:
an*

Reply:
resolved in 2ba646d118fe74b92c96890a26efea1e832fd404